metrics/internal/stackdriver/secret_watcher_test.go

```diff
@@ -0,0 +1,569 @@
+				}
+				return true, secrets, nil
+			}
+
+			return false, nil, nil
+		})
+
+		// Intercept all delete actions by the fake kubeclient and modify the return.
+		// Delete is used by SharedInformers for detecting OnDelete notifications.
+		f.Fake.PrependReactor("delete", "secrets", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
+			act := action.(k8stesting.DeleteAction)
+			ns := act.GetNamespace()
+			name := act.GetName()
+
+			// Omit "apple/one"
+			if ns == nsApple && name == nameOne {
+				return true, nil, nil
+			}
+
+			return false, nil, nil
+		})
+
+		testSetupWithCustomKubeclient(f)
+	}
+
+	tests := []struct {
+		name                 string
+		testSetupFunc        func()
+		watcherConstructor   func(observers ...Observer) SecretWatcher
+		secretToCreate       *secretForTest
+		shouldTriggerWatcher bool
+	}{
+		{
+			name:                 "GlobalWatcherChangeOrangeOne",
+			watcherConstructor:   globalSecretWatcherConstructor,
+			secretToCreate:       sOrangeOne,
+			shouldTriggerWatcher: true,
+		}, {
+			name:                 "GlobalWatcherChangeAppleOne",
+			watcherConstructor:   globalSecretWatcherConstructor,
+			secretToCreate:       sAppleOne,
+			shouldTriggerWatcher: true,
+		}, {
+			name:                 "GlobalWatcherChangeAppleTwo",
+			watcherConstructor:   globalSecretWatcherConstructor,
+			secretToCreate:       sAppleTwo,
+			shouldTriggerWatcher: true,
+		}, {
+			name:                 "AppleNamespaceWatcherChangeOrangeOne",
+			watcherConstructor:   appleNamespaceWatcherConstructor,
+			secretToCreate:       sOrangeOne,
+			shouldTriggerWatcher: false,
+		}, {
+			name:                 "AppleNamespaceWatcherChangeAppleOne",
+			watcherConstructor:   appleNamespaceWatcherConstructor,
+			secretToCreate:       sAppleOne,
+			shouldTriggerWatcher: true,
+		}, {
+			name:                 "AppleNamespaceWatcherChangeAppleTwo",
+			watcherConstructor:   appleNamespaceWatcherConstructor,
+			secretToCreate:       sAppleTwo,
+			shouldTriggerWatcher: true,
+		}, {
+			name:                 "AppleNameTwoSecretWatcherChangeOrangeOne",
+			watcherConstructor:   appleTwoSecretWatcherConstructor,
+			secretToCreate:       sOrangeOne,
+			shouldTriggerWatcher: false,
+		}, {
+			name:                 "AppleNameTwoSecretWatcherChangeAppleOne",
+			testSetupFunc:        appleNameTwoSecretWatcherChangeAppleOneTestSetup,
+			watcherConstructor:   appleTwoSecretWatcherConstructor,
+			secretToCreate:       sAppleOne,
+			shouldTriggerWatcher: false,
+		}, {
+			name:                 "AppleNameTwoSecretWatcherChangeAppleTwo",
+			watcherConstructor:   appleTwoSecretWatcherConstructor,
+			secretToCreate:       sAppleTwo,
+			shouldTriggerWatcher: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if test.testSetupFunc != nil {
+				test.testSetupFunc()
+			} else {
+				testSetup()
+			}
+
+			obsFuncs := &ObserverFuncs{
```

Comment:
What's the difference between this one and `defaultObserverFuncs`?

Reply:
Oh nothing, I'll use defaultObserverFuncs.